src/metabase/driver/mysql.clj

```diff
@@ -673,3 +677,132 @@
           (qp.writeback/execute-write-sql! db-id sql))
         (finally
           (.delete temp-file))))))
+
+(defn- parse-grant
+  "Parses the contents of a row from the output of a `SHOW GRANTS` statement, to extract the data needed
+   to reconstruct the set of table privileges that the current user has. Returns nil if the grant doesn't
+   contain any information we care about. Running `help show grants` in the mysql console shows the
+   syntax for the output strings of `SHOW GRANTS` statements.
+
+   There are two types of grants we care about: privileges and roles.
+
+   Privilege example:
+   (parse-grant \"GRANT SELECT, INSERT, UPDATE, DELETE ON `test-data`.* TO 'metabase'@'localhost'\")
+   =>
+   {:type       ::privileges
+    :privileges #{:select :insert :update :delete}
+    :object-type :table
+    :object-name \"test-data\"}
+
+   Role example:
+   (parse-grant \"GRANT 'example_role_1'@'%','example_role_2'@'%' TO 'metabase'@'localhost'\")
+   =>
+   {:type  ::roles
+    :roles #{'example_role_1'@'%' 'example_role_2'@'%'}}"
+  [grant]
+  (condp re-find grant
+    #"^GRANT PROXY ON "
+    nil
+    #"^GRANT (.+) ON FUNCTION "
+    nil
+    #"^GRANT (.+) ON PROCEDURE "
+    nil
+    ;; GRANT
+    ;;     priv_type [(column_list)]
+    ;;       [, priv_type [(column_list)]] ...
+    ;;     ON priv_level
+    ;;     TO user etc.
+    ;; }
+    ;; For now we ignore column-level privileges. But this is how we could get them in the future.
+    #"^GRANT (.+) ON (.+) TO "
+    :>>
+    (fn [[_ priv_types object_type]]
+      (when-let [privileges (not-empty
+                             (if (= priv_types "ALL PRIVILEGES")
+                               #{:select :update :delete :insert}
+                               (let [split-priv-types (->> (str/split priv_types #", ")
+                                                           (map (comp keyword u/lower-case-en)))]
+                                 (set/intersection #{:select :update :delete :insert} (set split-priv-types)))))]
+        {:type       ::privileges
+         :privileges privileges
+         :object-type (cond
+                        (= object_type "*.*")             :global
+                        (str/ends-with? object_type ".*") :database
+                        :else                             :table)
+         :object-name object_type}))
```

Comment:
> rename object_type and priv_types to object-type and priv_types
Seems like you made a typo. But anyway, it does seem like I was sloppy naming these keys. I decided to rename these in the following way:
- :privileges -> privilege-types (a privilege is something which contains an object and a privilege type)
- :object-type -> :level
- :object-name -> :object (name is misleading as this can be something like '*.*')

> add the (set ...) to the threading, i.e. (let [split-priv-types (->> (str/split ...) (map ...) (set)).

Yes, done

> I think it's a little clearer to call not-empty in the map (:privileges (not-empty privileges)) than wrapping the if.

Yes, done

> I'm not sure why ::privileges (and ::roles) is namespaced and :global, :database, and :table are not. I'd choose one or the other (probably not-namespaced).

Thanks for catching, I was experimenting with a convention where I imagined a global tagging system for maps, where a `type` key was always a namespaced keyword. But I'll shelve that idea for now and use not-namespaced.